components/sync15/src/telemetry.rs

```diff
@@ -246,17 +246,46 @@ pub enum SyncFailure {
     Unexpected { error: String },
 
     #[serde(rename = "autherror")]
-    Auth { from: String },
+    Auth { from: &'static str },
 
     #[serde(rename = "httperror")]
-    Http { code: u32 },
+    Http { code: u16 },
 }
 
-pub fn sync_failure_from_error(e: &Error) -> SyncFailure {
-    SyncFailure::Unexpected {
-        // TODO: Distinguish between error types, truncate, and anonymize
-        // messages.
-        error: e.to_string(),
+impl<'a> From<&'a Error> for SyncFailure {
+    fn from(e: &Error) -> SyncFailure {
+        match e.kind() {
+            ErrorKind::TokenserverHttpError(status) => {
+                if *status == 401 {
+                    SyncFailure::Auth {
+                        from: "tokenserver",
+                    }
+                } else {
+                    SyncFailure::Http { code: *status }
+                }
+            }
+            ErrorKind::BackoffError(_) => SyncFailure::Http { code: 503 },
+            ErrorKind::StorageHttpError(ref e) => match e {
+                ErrorResponse::NotFound { .. } => SyncFailure::Http { code: 404 },
+                ErrorResponse::Unauthorized { .. } => SyncFailure::Auth { from: "storage" },
+                ErrorResponse::PreconditionFailed { .. } => SyncFailure::Http { code: 412 },
+                ErrorResponse::ServerError { status, .. } => SyncFailure::Http { code: *status },
+                ErrorResponse::RequestFailed { status, .. } => SyncFailure::Http { code: *status },
+            },
+            ErrorKind::OpensslError(ref e) => SyncFailure::Unexpected {
+                error: e.to_string(),
+            },
+            ErrorKind::RequestError(ref e) => SyncFailure::Unexpected {
+                error: e.to_string(),
+            },
+            ErrorKind::UnexpectedStatus(ref e) => SyncFailure::Http { code: e.status },
+            ErrorKind::Interrupted(ref e) => SyncFailure::Unexpected {
+                error: e.to_string(),
+            },
+            e => SyncFailure::Other {
```

Comment:
Hmm, Wonder if it's worth doing anything about store errors, since they might have a nested sync15 error. Oh well, this is probably fine for now.

Reply:
Agreed, I decided to leave it alone for now, and see what kind of errors we get—we can unpack them later if we see too many in `unexpected` or `other`.